media-proxy/src/session-base.cc

```diff
@@ -0,0 +1,211 @@
+                       on_disconnect_callback_wrapper, on_receive_callback_wrapper, this);
+    if (ret != MEMIF_ERR_SUCCESS) {
+        INFO("memif_create: %s", memif_strerror(ret));
+        return -1;
+    }
+
+    /* Start the MemIF event loop. */
+    memif_event_thread = new std::thread(&session::memif_event_loop, this);
+    if (!memif_event_thread) {
+        ERROR("%s(%d), thread create fail\n", __func__, ret);
+        return -1;
+    }
+
+    return 0;
+}
+
+int session::shm_deinit()
+{
+    int err = 0;
+
+    if (memif_event_thread) {
+        memif_event_thread->join();
+        delete memif_event_thread;
+        memif_event_thread = 0;
+    }
+
+    /* free-up resources */
+    memif_delete(&memif_conn);
+    memif_delete_socket(&memif_socket);
+
+    /* unlink socket file */
+    if (memif_conn_args.is_master && memif_socket_args.path[0] != '@') {
+        unlink(memif_socket_args.path);
+    }
+
+    return 0;
+}
+
+int session::on_connect_cb(memif_conn_handle_t conn)
+{
+    INFO("RX RDMA memif connected!");
+
+    int err = memif_refill_queue(conn, 0, -1, 0);
+    if (err != MEMIF_ERR_SUCCESS) {
+        INFO("memif_refill_queue: %s", memif_strerror(err));
+        return err;
+    }
+
+    print_memif_details(conn);
+
+    atomic_store_explicit(&shm_ready, true, memory_order_release);
+
+    return 0;
+}
+
+int session::on_disconnect_cb(memif_conn_handle_t conn)
+{
+    if (!conn) {
+        INFO("Invalid Parameters.");
+        return -EINVAL;
+    }
+
+    /* release session */
+    if (!atomic_load_explicit(&shm_ready, memory_order_relaxed)) {
+        return 0;
+    }
+    atomic_store_explicit(&shm_ready, false, memory_order_relaxed);
+
+    /* stop event polling thread */
+    INFO("Stop poll event\n");
+    memif_socket_handle_t socket = memif_get_socket_handle(conn);
+    if (!socket) {
+        INFO("Invalide socket handle.");
+        return -1;
+    }
+
+    int err = memif_cancel_poll_event(socket);
+    if (err != MEMIF_ERR_SUCCESS) {
+        INFO("We are doomed...");
+    }
+
+    return 0;
+}
+
+int session::on_receive_cb(memif_conn_handle_t conn, uint16_t qid)
+{
+    memif_buffer_t shm_bufs;
+    uint16_t buf_num = 0;
+
+    /* receive packets from the shared memory */
+    int err = memif_rx_burst(conn, qid, &shm_bufs, 1, &buf_num);
+    if (err != MEMIF_ERR_SUCCESS && err != MEMIF_ERR_NOBUF) {
+        INFO("memif_rx_burst: %s", memif_strerror(err));
+        return err;
+    }
+
+    /* Process on the received buffer. */
+    /* Supposed this function will never be called. */
+
+    err = memif_refill_queue(conn, qid, buf_num, 0);
+    if (err != MEMIF_ERR_SUCCESS)
+        INFO("memif_refill_queue: %s", memif_strerror(err));
+
+    return 0;
+}
+
+session::session() : id(0), memif_socket(0), memif_conn(0)
+{
+    memif_event_thread = 0;
+    type = TX;
+    payload_type = PAYLOAD_TYPE_NONE;
+    memif_socket_args = {0};
+    memif_conn_args = {0};
+}
```

Comment:
What's the reason to split initialization of the fields into the two parts, the function header and the function body?

Reply:
fixed